src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -137,13 +137,39 @@ public override BoundNode VisitCall(BoundCall node)
                     bool mightMutate =
                         // might be a call to a local function that assigns something
                         node.Method.MethodKind == MethodKind.LocalFunction ||
-                        // or perhaps we are passing a variable by ref and mutating it that way
-                        !node.ArgumentRefKindsOpt.IsDefault;
+                        // or perhaps we are passing a variable by ref and mutating it that way, e.g. `int.Parse(..., out x)`
+                        !node.ArgumentRefKindsOpt.IsDefault ||
+                        // or perhaps we are calling a mutating method of a value type
+                        MethodMayMutateReceiver(node.ReceiverOpt, node.Method);
 
                     if (mightMutate)
                         _mightAssignSomething = true;
                     else
                         base.VisitCall(node);
+                    return null;
+                }
+
+                private static bool MethodMayMutateReceiver(BoundExpression receiver, MethodSymbol method)
+                {
+                    return
+                        method != null &&
+                        !method.IsStatic &&
+                        !method.IsEffectivelyReadOnly &&
+                        receiver.Type?.IsReferenceType == false &&
+                        // methods of concrete special types do not mutate their receiver
+                        (method.ContainingType.SpecialType == SpecialType.None || method.ContainingType.TypeKind == TypeKind.Interface);
```

Comment:
>method.ContainingType.SpecialType == SpecialType.None [](start = 25, length = 53)

Please check for specific types that we know about, the set of Special types is changing and we should not assume that all of them will be like that. I think there is a way to check if a type is primitive, perhaps that be sufficient for the purpose of this change.